File: tests/fixtures/generic/clean/taint_open_redirect_clean.cs
// @audit-fixture
// @rule: taint_open_redirect
// @category: security
// @expected: clean
public class AccountController : Controller {
    public IActionResult Login() {
        var returnUrl = Request.Query["returnUrl"].ToString();
        if (!Url.IsLocalUrl(returnUrl)) returnUrl = "/";
        return Redirect(returnUrl);
    }
}

File: tests/fixtures/generic/clean/taint_open_redirect_clean.js
// @audit-fixture
// @rule: taint_open_redirect
// @category: security
// @expected: clean
const express = require("express");
const app = express();
const ALLOWED = ["example.com"];
app.get("/login", (req, res) => {
    const next = req.query.next;
    const url = new URL(next);
    if (!ALLOWED.includes(url.host)) return res.redirect("/");
    res.redirect(url.toString());
});
